### PR TITLE
Fix lock contention issue scanning statics

### PR DESCRIPTION
**Pull request by PeterSolMS**
This is the fix to the lock contention issue while scanning statics in background_mark_phase.

The basic idea is to gather the GC statics from the first thread done with GCScan::GCScanRoots() and finalize_queue->GcScanRoots(…). Threads arriving later can just bulk-copy the roots. This eliminates the lock contention, but still gives us work distribution across threads.

During the synchronous phase when we catch up to what the application threads have been doing, the scanning of the statics is also done on the first thread that is done with its other work. We also get rid of the lock contention here, but do not get work distribution. This should be ok, as there shouldn't be much work to do, and any changed statics should also get caught by revisit_written_pages.

I added a flag "scanStatics" to the ScanContext struct to control the behavior of GCScan::GCScanRoots().

Unsolved issue: what to do about blocking gen 2 collections. My change will eliminate the lock contention for scanning statics for blocking gen 2 collections, but we will also lose some work balancing we were getting before. We could do the work on the first thread arriving like I did in background_mark_phase, but I'm not sure it's the best solution.

### CODE REVIEW

**Comment by VSadov** [comment]
I am curious:
- Could this just replace lock contention with join contention?
- what is protected by the lock? (i can't easily find when it is taken/released)

**Comment by PeterSolMS** [comment]
This does replace a large number of lock acquisitions (#modules*#heaps) with a much smaller number of joins (#heaps). So it's a win even with a modest number of modules.

The lock protects the assembly list. It's acquired and released here:

```
BOOL
AppDomain::AssemblyIterator::Next(
    CollectibleAssemblyHolder<DomainAssembly *> * pDomainAssemblyHolder)
{
...
   CrstHolder ch(m_pAppDomain->GetAssemblyListLock());
    return Next_Unlocked(pDomainAssemblyHolder);
}
```

To be sure, none of the GC threads will actually change this list, and all the managed threads should be suspended, So you'd think that we're safe even without a lock, but there could be a race condition where a managed thread enters the runtime right before the suspension, becomes preemptive, and proceeds to manipulate this list right as we are traversing it on the GC threads.

**Comment by jkotas** [comment]
Have you looked at the option to always manipulate this list in cooperative mode? How would the behavior compare to this change?

**Comment by PeterSolMS** [comment]
I thought about it, but worried that I didn't understand the implications well enough to be confident of correctness. For instance we also call the assembly iteration from regular (i.e. non-background) GC threads - does cooperative or preemptive mode even have meaning for these, or would some code in the runtime crash when it tried to force cooperative mode?

This solution would solve the performance issue about as well though.

File: src/coreclr/src/gc/gcinterface.h [hunk 1]
```diff
@@ -942,6 +942,7 @@ struct ScanContext
     uintptr_t stack_limit; // Lowest point on the thread stack that the scanning logic is permitted to read
     bool promotion; //TRUE: Promotion, FALSE: Relocation.
     bool concurrent; //TRUE: concurrent scanning
+    bool scanStatics; //TRUE: scan statics, not stack
```
Comment (jkotas) [inline src/coreclr/src/gc/gcinterface.h]: The contract for this looks odd: The GC can set it, but then the EE side can decide that it is not a good idea to scan the statics and ignore it. 

Should this rather be an argument for `GCToEEInterface::GcScanRoots` and have the policy to decide whether or not to scan statics on the GC side? Or maybe even have a new `GCToEEInterface::GcScanStatics` method on GC/EE interface?

**Comment by VSadov** [comment]
Hmm. This is a bit unfortunate. 
What I see is that GC threads do not need mutual exclusion among themselves when accessing the list (for correctness), but this change introduces such guarantee to make inner locks cheaper. Also it seems the list is modified very rarely. 

It does look like changing the iterators in such a way that pure readers would not need locks is not trivial.

The code also mentions that this scan is optional since all these objects are also rooted via handles.
I am wondering - what happens if we just do not scan statics here?

**Comment by Maoni0** [comment]
scanning statics this way was specifically added for perf and statics were already reachable via handles. so I would presume there's merit that warranted such a change. if it's not much work/not risky to keep such a change I'd prefer to keep it. so far our conclusion was it's not much work/not risky so that's why we went for a fix.

**Comment by VSadov** [comment]
I am mostly curious where the perf advantages come from. It is not obvious when you eagerly do something that is optional.
I also wonder whether gains are big enough - to make sure that adding a join and mark list copying would not negate those gains. 

I assume Peter has a way to measure this scenario, so it makes sense to measure with the whole deal disabled. At least as a data point - to know what there is profit.

**Comment by Maoni0** [comment]
the gains can be very significant. essentially this is a way to have multiple threads to pick object graphs that can be very large. so the 1st thread will pick one, mark it as in progress which means the 2nd thread will leave that one alone and pick the next one. this was the intent of the code when it was added, as I understood.

**Comment by jkotas** [comment]
> the gains can be very significant. essentially this is a way to have multiple threads to pick object graphs

Is it possible to build a micro-benchmark that will regress significantly when the one-off statics scanning is deleted? We should have this micro-benchmark in the suite of GC perf tests.

**Comment by Maoni0** [comment]
> Is it possible to build a micro-benchmark that will regress significantly when the one-off statics scanning is deleted? We should have this micro-benchmark in the suite of GC perf tests.

yes, it takes some work. I'll talk to Peter about it.

**Comment by PeterSolMS** [comment]
I have found another solution that I believe is better and submitted another pull request for it.

**Comment by VSadov** [comment]
Since there is a different better solution, perhaps this PR can be closed?

**Pull request closed without merging**